web/src/settings_components.ts

```diff
@@ -31,24 +48,60 @@ export function get_sorted_options_list(option_values_object) {
             }
             return 0;
         };
+    } else {
+        comparator = (x, y) => x.order - y.order;
     }
     options_list.sort(comparator);
     return options_list;
 }
 
-export function get_realm_time_limits_in_minutes(property) {
-    if (realm[property] === null) {
-        // This represents "Anytime" case.
-        return null;
-    }
-    let val = (realm[property] / 60).toFixed(1);
-    if (Number.parseFloat(val, 10) === Number.parseInt(val, 10)) {
-        val = Number.parseInt(val, 10);
-    }
-    return val.toString();
-}
+type message_time_limit_settings =
+    | "realm_message_content_edit_limit_seconds"
+    | "realm_move_messages_between_streams_limit_seconds"
+    | "realm_move_messages_within_stream_limit_seconds"
+    | "realm_message_content_delete_limit_seconds";
 
-export function get_property_value(property_name, for_realm_default_settings, sub, group) {
+export function get_realm_time_limits_in_minutes(property: message_time_limit_settings): string {
+    const setting_value = realm[property];
+    if (setting_value === null) {
+        // This represents "Anytime" case.
+        return "";
+    }
+    let val = (setting_value / 60).toFixed(1);
+    if (Number.parseFloat(val) === Number.parseInt(val, 10)) {
+        val = (setting_value / 60).toFixed(0);
+    }
+    return val;
+}
+
+type RealmSetting = Omit<typeof realm, "realm_authentication_methods"> & {
+    realm_authentication_methods: Record<string, boolean>;
+};
+type RealmSettingProperties = keyof RealmSetting | "realm_org_join_restrictions";
+
+type RealmUserSettingDefaultType = typeof realm_user_settings_defaults;
+type RealmUserSettingDefaultProperties =
+    | keyof RealmUserSettingDefaultType
+    | "email_notification_batching_period_edit_minutes";
+
+type StreamSettingProperties = keyof StreamSubscription | "stream_privacy" | "is_default_stream";
+
+type valueof<T> = T[keyof T];
+
+export function get_property_value(
+    property_name:
+        | RealmSettingProperties
+        | StreamSettingProperties
+        | keyof UserGroup
+        | RealmUserSettingDefaultProperties,
+    for_realm_default_settings?: boolean,
+    sub?: StreamSubscription,
+    group?: UserGroup,
+):
+    | valueof<RealmSetting>
+    | valueof<StreamSubscription>
+    | valueof<UserGroup>
+    | valueof<RealmUserSettingDefaultType> {
```

Comment:
This is not a suitable use case for unions. Use generics to get a precise return type. (I can try to provide more detailed help later.)

Reply:
I tried doing this but was not able to do it correctly. I did something like 
```js
export function get_property_value<T>(
    property_name: keyof T,
    for_realm_default_settings?: boolean,
    sub?: StreamSubscription,
    group?: UserGroup,
): valueof<T> {
```

Maybe we would want it like this to make sure that T is of correct type
```js
type setting_type = typeof realm | typeof realm_user_settings_defaults | StreamSubscription | UserGroup;
export function get_property_value<T extends setting_type>(
    property_name: keyof T,
    for_realm_default_settings?: boolean,
    sub?: StreamSubscription,
    group?: UserGroup,
): valueof<T> {
```

Reply:
`valueof<T>` is not a useful construct, since it gives unions that we don’t want. This function should be generic over **the string literal type of `property_name`**, and should return exactly the type of that one property. TypeScript is powerful enough to support this. Look at [`getConfigItem` in zulip-desktop](https://github.com/zulip/zulip-desktop/blob/v5.10.5/app/common/config-util.ts#L26) for an example.

Reply:
I tried doing it like this -
```js
type Setting = typeof realm | typeof realm_user_settings_defaults | StreamSubscription | UserGroup;
export function get_property_value<Key extends keyof Setting>(
    property_name: Key,
    for_realm_default_settings?: boolean,
    sub?: StreamSubscription,
    group?: UserGroup,
): Setting[Key] {
```

But then the typescript gives error like `Argument of type 'string' is not assignable to parameter of type 'never'` at the lines where `get_property_value` function is called. I am not sure how the `never` type comes into picture here.

Reply:
Maybe due to the compiler not being sure the key passed will be present in the object, which will be decided based on other parameters passed to the function.